src/stash.rs

```diff
@@ -0,0 +1,287 @@
+use {raw, panic, StashFlags, Oid, Error, Repository, Signature, StashApplyProgress};
+use std::ffi::{CString, CStr};
+use util::{Binding};
+use libc::{c_int, c_char, size_t, c_void, c_uint};
+use build::{CheckoutBuilder};
+use std::mem;
+
+/// Stash application progress notification function.
+///
+/// Return `true` to continue processing, or `false` to
+/// abort the stash application.
+pub type StashApplyProgressCb<'a> = FnMut(StashApplyProgress) -> bool + 'a;
+
+/// This is a callback function you can provide to iterate over all the
+/// stashed states that will be invoked per entry.
+pub type StashCb<'a> = FnMut(usize, &str, &Oid) -> bool + 'a;
+
+#[allow(unused)]
+/// Stash application options structure
+pub struct StashApplyOptions<'cb> {
+    progress: Option<Box<StashApplyProgressCb<'cb>>>,
+    raw_opts: raw::git_stash_apply_options
+}
+
+impl<'cb> StashApplyOptions<'cb> {
+    /// Creates a default set of merge options.
+    pub fn new() -> StashApplyOptions<'cb> {
+        let mut opts = StashApplyOptions {
+            progress: None,
+            raw_opts: unsafe { mem::zeroed() },
+        };
+        assert_eq!(unsafe {
+            raw::git_stash_apply_init_options(&mut opts.raw_opts, 1)
+        }, 0);
+        opts
+    }
+
+    /// Set stash application flag
+    pub fn flags(&mut self, flags: raw::git_stash_apply_flags) -> &mut StashApplyOptions<'cb> {
+        self.raw_opts.flags = flags;
+        self
+    }
+
+    /// Options to use when writing files to the working directory
+    pub fn checkout_options(&mut self, mut opts: CheckoutBuilder) -> &mut StashApplyOptions<'cb> {
+        unsafe {
+            opts.configure(&mut self.raw_opts.checkout_options);
+        }
+        self
+    }
+
+    /// Optional callback to notify the consumer of application progress.
+    ///
+    /// Return `true` to continue processing, or `false` to
+    /// abort the stash application.
+    pub fn progress_cb(&mut self, callback: Box<StashApplyProgressCb<'cb>>) -> &mut StashApplyOptions<'cb> {
+        self.progress = Some(callback);
+        self.raw_opts.progress_cb = stash_apply_progress_cb;
+        self.raw_opts.progress_payload = self as *mut _ as *mut _;
+        self
+    }
+
+    /// Pointer to a raw git_stash_apply_options
+    fn raw(&self) -> &raw::git_stash_apply_options {
+        &self.raw_opts
+    }
+}
+
+/// A set of stash methods
+pub struct Stash {
+
+}
+
+impl Stash {
+    /// Save the local modifications to a new stash.
+    pub fn save(repo: &mut Repository,
```

Comment:
I've typically preferred to add these sorts of methods onto `Repository` itself, so we'd have something like:

```rust
repo.stash_save(...);
repo.stash_apply(...);
// etc
```